Snowflake.Data/Core/FileTransfer/FileBackedOutputStream.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            _outputStream.SetLength(value);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            SwitchToFileIfTooMuchInMemory(count);
+            _outputStream.Write(buffer, offset, count);
+        }
+
+        public override bool CanRead => _outputStream.CanRead;
+        public override bool CanSeek => _outputStream.CanSeek;
+        public override bool CanWrite => _outputStream.CanWrite;
+        public override long Length => _outputStream.Length;
+        public override long Position
+        {
+            get => _outputStream.Position;
+            set => _outputStream.Position = value;
+        }
+        
+        [Obsolete("It is not a good idea to load large stream into memory at once. Created only for backward compatibility issues.")]
+        public byte[] ToArray()
+        {
+            if (IsUsingFileOutputStream())
+            {
+                using (var memoryStream = new MemoryStream())
+                {
+                    var originalPosition = _fileOutputStream.Position;
+                    _fileOutputStream.Position = 0;
+                    _fileOutputStream.CopyTo(memoryStream);
+                    _fileOutputStream.Position = originalPosition;
+                    return memoryStream.ToArray();
+                }
+            }
+            else
+            {
+                return _memoryOutputStream.ToArray();
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _outputStream.Close();
+            base.Dispose(disposing);
+        }
+
+        internal String GetFileName()
+        {
+            return _fileName;
+        }
+
+        private void SwitchToFileIfTooMuchInMemory(int numberOfBytesToConsume)
+        {
+            if (IsUsingFileOutputStream())
+            {
+                return;
+            }
+            if (Length + numberOfBytesToConsume <= _maxInMemoryBytes)
+            {
+                return;
+            }
+            SwitchFromMemoryToTempFile();
+        }
+
+        internal bool IsUsingFileOutputStream()
+        {
+            return _fileOutputStream != null;
+        }
+        
+        private void SwitchFromMemoryToTempFile()
+        {
+            _fileName = GenerateTempFilePath();
+            _fileOutputStream = new FileStream(_fileName, FileMode.Create, FileAccess.ReadWrite, FileShare.None, 4096,
+                FileOptions.DeleteOnClose);
+            _memoryOutputStream.Position = 0;
+            _memoryOutputStream.CopyTo(_fileOutputStream);
+            _outputStream = _fileOutputStream;
+            _memoryOutputStream.Close();
+            _memoryOutputStream = null;
+        }
+
+        private string GenerateTempFilePath()
+        {
+            var millisFromEpoch = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+            var randomFileName = $"FileBackedOutputStream_{millisFromEpoch}.tmp";
```

Comment:
no risk related  to parallel processing?  and having accidentally two files with same  timestamp?

Reply:
normally the file is created inside a random folder dedicated to given file uploading/downloading process.
For example let it be `/tmp/abcd.efg` folder, where `/tmp` is a default temp folder, and `abcd.efg` is a random directory name (despite it looks like a file name). So when you would like to upload 2 files at the exactly the same time it the files would be created in different folders. However, to be more certain about the solution, I added a randomized part to the file name.